test/gpu/native/studies/shoc/sort.chpl

```diff
@@ -0,0 +1,679 @@
+                    ref blockOffsets : [] uint(32),
+                    ref scanBlockSums : [] list(uint(32)), // scanBlockSums is an array of lists
+                    numElements: uint(32)){
+
+    // Threads have either 4 or 2 elements each
+    const radixGlobalWorkSize   = numElements / 4;
+    const findGlobalWorkSize    = numElements / 2;
+    const reorderGlobalWorkSize = numElements / 2;
+
+    // Radix Kernel uses block size of 128, others use 256
+    // (same as scan)
+    const radixBlocks   = radixGlobalWorkSize   / SORT_BLOCK_SIZE;
+    const findBlocks    = findGlobalWorkSize    / SCAN_BLOCK_SIZE;
+    const reorderBlocks = reorderGlobalWorkSize / SCAN_BLOCK_SIZE;
+
+    // radixSortBlocks Kernel here
+    radixSortBlocks(radixGlobalWorkSize, nbits, startbit, tempKeys, tempValues, keys, values);
+
+    // findRadixOffset Kernel here
+    // Use offsets again
+    findRadixOffsets(findGlobalWorkSize, tempKeys, counters, blockOffsets, startbit, numElements, findBlocks);
+
+    scanArrayRecursive(countersSum, counters, 16*reorderBlocks, 0, scanBlockSums);
+
+    // reorderData Kernel here
+    reorderData(reorderGlobalWorkSize, startbit, keys, values, tempKeys, tempValues,
+         blockOffsets, countersSum, counters, reorderBlocks);
+}
+
+proc scanArrayRecursive(ref outArray : [] uint(32), ref inArray : [] uint(32),
+                        numElements, level, ref blockSums : [] list(uint(32))){
+    // Kernels hadle 8 elements per thread
+    const numBlocks : uint(32) = max(1,ceil(numElements:real
+        / (4.0:real * SCAN_BLOCK_SIZE)) : uint(32)) : uint(32);
+
+    const sharedEltsPerBlock : uint(32) = (SCAN_BLOCK_SIZE * 2) : uint(32);
+    const sharedMemSize : uint(32) = numBytes(uint(32)) * sharedEltsPerBlock;
+
+    const fullBlock : bool = (numElements == numBlocks * 4 * SCAN_BLOCK_SIZE);
+
+    // CUDA version uses dim3 for no apparent reason because they
+    // created dim3 with only x populated for grid and block size
+
+    //execute scan
+    if(numBlocks>1){
+        // Scan Kernel here
+        var t = blockSums[level].toArray();
+        scanKernel(numBlocks, outArray, inArray, t, numElements, fullBlock, true);
+        blockSums[level] = new list(t);
+    } else {
+        // Different Scan Kernel here
+        var t = blockSums[level].toArray();
+        scanKernel(numBlocks, outArray, inArray, t, numElements, fullBlock, false);
+        blockSums[level] = new list(t);
+    }
+
+    if(numBlocks > 1){
+        var t = blockSums[level].toArray();
+        scanArrayRecursive(t, t, numBlocks, level+1, blockSums);
+        blockSums[level] = new list(t);
+        // VectorAddUniform4 Kernel Here
+        t = blockSums[level].toArray();
+        vectorAddUniform4(outArray, t, numElements);
+        blockSums[level] = new list(t);
+    }
+}
+
+proc verifySort(ref keys : [] uint(32), ref values : [] uint(32), const size){
+    var passed = true;
+    for i in 0..<size-1 {
+        if(keys[i] > keys[i+1]){
```

Comment:
Do you know if the end result of this is that we sort based on keys or values?

I guess the program sets `keys = values` and does the sort based on keys?

I find that a little surprising. I would have guessed that keys were the original indexes into the array and values were the values we were sorting.

Anyway I can see you're copying what the CUDA version does, just curious if you have a theory for why they do it this way (I guess they're trying to simulate sorting a map or something?).

Reply:
Yeah they are trying to simulate sorting a map.
They sort based on the keys and only check to make sure that they keys are sorted. 
They do not check to make sure the key-value pairs are the same which is kind of weird to me. 

If it were me I would also have the `verifySort` function check if `key[i] == value[i]`.